### PR TITLE
tsdb: Bug fix for further continued deletions after crash deletions; added more tests.

### PR DESCRIPTION
**Pull request by bwplotka**
Additionally: Added log line for block removal.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by brancz** [comment]
#7777 jackpot

lgtm

**Comment by roidelapluie** [comment]
> #7777 jackpot
> 
> lgtm, nice!

In this PR everything is permitted :)

**Comment by bwplotka** [comment]
Let me commit remote write receiver then.... ;p

File: tsdb/db.go [hunk 1]
```diff
@@ -1135,14 +1135,23 @@ func (db *DB) deleteBlocks(blocks map[ulid.ULID]*Block) error {
 			}
 		}
 
-		// Replace atomically to avoid partial block when process is crashing during this moment.
+		toDelete := filepath.Join(db.dir, ulid.String())
+		if _, err := os.Stat(toDelete); os.IsNotExist(err) {
```
Comment (kakkoyun) [inline tsdb/db.go:1139]: 👍

File: tsdb/db.go [hunk 2]
```diff
@@ -921,20 +921,23 @@ func (db *DB) reload() (err error) {
...
+	// crashing towards the end of a compaction but before deletions. By doing that, we can pick up the deletion where it left off during a crash.
 	for _, block := range loadable {
 		if _, ok := deletableULIDs[block.meta.ULID]; ok {
 			deletable[block.meta.ULID] = block
 		}
 		for _, b := range block.Meta().Compaction.Parents {
-			delete(corrupted, b.ULID)
+			if _, ok := corrupted[b.ULID]; ok {
+				delete(corrupted, b.ULID)
+				level.Warn(db.logger).Log("msg", "Found corrupted block, but replaced by compacted one so it's safe to delete. This should not happen with atomic deletes. Please Report", "block", b.ULID)
```
Comment (roidelapluie) [inline tsdb/db.go:933]: That seems strange.

We never do this.

Where should they report it to?
santa

We are not the sole distributors of prometheus and sometimes it is even modified (e.g. debian package) so I would avoid this or find a general rule.

Comment (bwplotka) [reply tsdb/db.go:933]: Makes sense (:

File: tsdb/db.go [hunk 3]
```diff
@@ -1135,14 +1135,23 @@ func (db *DB) deleteBlocks(blocks map[ulid.ULID]*Block) error {
...
+		// Replace atomically to avoid partial block when process would crash during deletion.
 		tmpToDelete := filepath.Join(db.dir, fmt.Sprintf("%s%s", ulid, tmpForDeletionBlockDirSuffix))
-		if err := fileutil.Replace(filepath.Join(db.dir, ulid.String()), tmpToDelete); err != nil {
+		if err := fileutil.Replace(toDelete, tmpToDelete); err != nil {
 			return errors.Wrapf(err, "replace of obsolete block for deletion %s", ulid)
 		}
 		if err := os.RemoveAll(tmpToDelete); err != nil {
 			return errors.Wrapf(err, "delete obsolete block %s", ulid)
 		}
+		level.Info(db.logger).Log("msg", "Deleting obsolete block", "block", ulid)
```
Comment (codesome) [inline tsdb/db.go:1154]: Late for review, but should this be `Deleted`? :)

Comment (bwplotka) [reply tsdb/db.go:1154]: Will fix in separare PR, thanks!

**Pull request merged**